tensorflow/core/platform/posix/posix_file_system.cc

```diff
@@ -276,4 +281,57 @@ Status PosixFileSystem::RenameFile(const string& src, const string& target) {
   return result;
 }
 
+Status PosixFileSystem::CopyFile(const string& src, const string& target) {
+  string translated_src = TranslateName(src);
+  struct stat sbuf;
+  if (stat(translated_src.c_str(), &sbuf) != 0) {
+    return IOError(src, errno);
+  }
+  int src_fd = open(translated_src.c_str(), O_RDONLY);
+  if (src_fd < 0) {
+    return IOError(src, errno);
+  }
+  string translated_target = TranslateName(target);
+  // O_WRONLY | O_CREAT:
+  //   Open file for write and if file does not exist, create the file.
+  // S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH:
+  //   Create the file with permission of 0666
+  int target_fd =
+      open(translated_target.c_str(), O_WRONLY | O_CREAT,
+           S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH);
+  if (target_fd < 0) {
+    close(src_fd);
+    return IOError(target, errno);
+  }
+  int rc = 0;
+  off_t offset = 0;
+  while (offset < sbuf.st_size) {
+    size_t chunk = (sbuf.st_size - offset) < SSIZE_MAX ? (sbuf.st_size - offset)
+                                                       : SSIZE_MAX;
+#if !defined(__APPLE__)
```

Comment:
Having done more homework, I would say change this to `#if defined(__linux__) && !defined(__ANDROID__)` because it won't work on FreeBSD and you never know with Android. I've confirmed it will work on Linux kernels ≥2.6.33 (published in 2010.) TensorFlow supports as far back as CentOS 7, which uses Linux 3.10.0, so this should be fine. If we want to be even more conservative, we could use the [splice](https://en.wikipedia.org/wiki/Splice_(system_call)) system call instead, which became available in 2.6.17 in 2006. It's Linux-only, but we've constrained ourselves to Linux anyway.

Reply:
Thanks @jart. Done.